src/xc_integrator/xc_data/device/xc_device_data.hpp

```diff
@@ -392,6 +459,11 @@ struct XCDeviceData {
    */
   virtual void retrieve_exc_vxc_integrands( double* EXC, double* N_EL,
```

Comment:
Similar to send_density, this needs to be a single, nullptr-safe function to avoid code dup